### PR TITLE
Elf version

### PR DESCRIPTION
**Pull request by calixteman**
Need to add some tests, but it should work.
The results are quite different from what we've with breakpad/dump_syms mainly because of the way to un-inline inlinees is better.

### CODE REVIEW

**Comment by froydnj** [comment]
I am happy to review this, but I'm also not going to get around to reviewing this until May because I am on leave.

> The results are quite different from what we've with breakpad/dump_syms mainly because of the way to un-inline inlinees is better.

Do those changes make stack traces with inlined functions more or less understandable than the breakpad-generated symbol files?

**Comment by calixteman** [comment]
I'd say that it's a way better since the lines belonging to an inlinee called at line A will appear in the inliner at line A and the line with same line (& file of course) are merged.
So normally we shouldn't see inlinees anymore.

File: Cargo.toml [hunk 1]
```diff
@@ -24,6 +24,10 @@ symbolic-common = "7.2"
 symbolic-debuginfo = "7.2"
 symbolic-demangle = "7.2"
 symbolic-minidump = "7.2"
+#symbolic-common = { path = "../symbolic.calixteman/common" }
+#symbolic-debuginfo = { path = "../symbolic.calixteman/debuginfo" }
+#symbolic-demangle = { path = "../symbolic.calixteman/demangle" }
+#symbolic-minidump = { path = "../symbolic.calixteman/minidump" }
```
Comment (nnethercote) [inline Cargo.toml:30]: ?

File: Cargo.toml [hunk 2]
```diff
@@ -38,3 +42,10 @@ vendored-openssl = ["openssl/vendored"]
 [dependencies.openssl]
 version = "0.10"
 optional = true
+
+[patch.crates-io]
+gimli = { git = "https://github.com/gimli-rs/gimli/", rev = "53c802a69066b211984ef06fd96ba20946d47ca3"}
+symbolic-common = { git = "https://github.com/calixteman/symbolic/", rev = "79138982daebf3654c18f6b37ea554636f4f7ca8" }
+symbolic-debuginfo = { git = "https://github.com/calixteman/symbolic/", rev = "79138982daebf3654c18f6b37ea554636f4f7ca8" }
+symbolic-demangle = { git = "https://github.com/calixteman/symbolic/", rev = "79138982daebf3654c18f6b37ea554636f4f7ca8" }
+symbolic-minidump = { git = "https://github.com/calixteman/symbolic/", rev = "79138982daebf3654c18f6b37ea554636f4f7ca8" }
```
Comment (nnethercote) [inline Cargo.toml:51]: Why are these necessary? Comments would be helpful.

File: src/line.rs [hunk 3]
```diff
@@ -111,6 +106,15 @@ impl Lines {
         last.len = sym_len - (last.rva - sym_rva);
     }
 
+    pub(crate) fn pdb_finalize(&mut self, sym_rva: u32, sym_len: u32, address_map: &AddressMap) {
+        self.compute_len(sym_rva, sym_len);
+        self.compute_rva(address_map);
+    }
+
+    pub(crate) fn elf_finalize(&mut self, sym_rva: u32, sym_len: u32) {
```
Comment (nnethercote) [inline src/line.rs:114]: Does it make sense for `pdb_finalize` and `elf_finalize` to instead be implemented in a `finalize` method in the `Dumpable` trait?

File: src/linux/source.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    ref_to_id: HashMap<PathBuf, u32>,
+    fake_id_to_ref: Vec<(Option<u32>, String)>,
+    id_to_ref: Vec<String>,
+    cache: HashMap<(SliceRef, SliceRef, SliceRef), u32>,
+}
+
+impl SourceFiles {
+    #[inline(always)]
+    fn cast_ptr(name: &[u8]) -> SliceRef {
+        unsafe { std::mem::transmute::<&[u8], SliceRef>(name) }
```
Comment (nnethercote) [inline src/linux/source.rs:25]: Using `as_ptr` and `len` would be less gross than using `transmute`.

File: src/line.rs (hunk 3 above)
Comment (gabrielesvelto) [reply src/line.rs:114]: Agreed, that would keep all the platform-specific bits in their appropriate sources.

File: Cargo.toml (hunk 1 above)
Comment (gabrielesvelto) [reply Cargo.toml:30]: Some of Calixte's PRs haven't been merged upstream yet. IIRC the changes are about how lines are handled (for inlined functions and the like). Originally we though we'd write that functionality in **dump_syms** like we currently have in the Breakpad version but the Sentry team expressed interest of having that into Symbolic and so Calixte is doing the work upstream.

File: src/action.rs [hunk 5]
```diff
@@ -78,14 +82,26 @@ impl Action<'_> {
...
                             windows::utils::try_to_set_pe(&path, &mut pdb, &buf);
-                            dumper.store_pdb(&pdb)
+                            dumper.store(&pdb)
                         }
                         Err(e) => Err(e.into()),
                     }
                 }
+                "dbg" | "so" => {
+                    let (buf, filename) = dumper.get_from_id(&path, filename, dumper.debug_id)?;
+                    let elf = linux::elf::ElfInfo::new(&buf, filename, true /* with_stack */)?;
```
[the remark this answers is not in the document]
Comment (calixteman) [reply src/action.rs]: I'll do it in a followup patch.
